### ISSUE TITLE
Spots(row) and celltypes (column) are getting dropped in the output cell type proportion matrix

### ISSUE THREAD

**Issue by ajitashree** [lzygenomics/SONAR#3]
Dear authors,

We are running your method SONAR for spot-deconvolution task. However, we are running into some issues post the following command in MATLAB.

**eval(['save ',thepath,'SONAR_results',' JIE'])**

Multiple spots were missing in the output cell type proportion matrix. Also, the number of cell types were reduced to 12 (17 cell types were originally present in the single-cell data). 

Please help us resolve the issue.

**Comment by lzygenomics** [comment]
Thank you for being interested in our algorithm. If only a few spots are missing, they may have been filtered out during the quality control process. However, a reduction in the number of cell types is unlikely to occur. I suspect there may have been a problem with the intermediate file transfer and generation. Could you please provide the scripts you used?

**Comment by Aditya-Venkat** [comment]
Hi, we thank you for the immediate response. The R script used (matlab script was unmodified) is given below.

Upon reading the output using ,
,SONAR.results <- readMat(paste(code_path,"SONAR_results.mat",sep = ""), drop = FALSE)
,we are able to notice a significant reduction in number of spots and a reduced number of cell types. The processed files delivered using SONAR.deliver have the expected number of cells and spots. 

We also tried changing the matlab code to generate a CSV file instead of a .mat file and recieved the correct number of cells and a loss of one spot in the output. We are unable to identify why the SONAR.results object is not having the desired number of spots and cell types.
```
# library(devtools)
# 
# devtools:: install_github("lzygenomics/SONAR")
options(matlab.path="/home/ajita/Desktop")
library(anndata)
library(SONAR)
# library(here)
library(DescTools)
library(scater)
library(Matrix)
library(data.table)
library(Seurat)
library(matlabr)
library(R.matlab)
# library(anndata)


i <- as.character(151508)
code_path <- "/data/ajita/Spatial/benchmarking_spatial_deconvolution/SONAR/core-code/"
#scrna data
setwd("/data/ajita/Spatial/Datasets/Spatial_Deconvolution/DLPFC")
SC <- read_h5ad("SC.h5ad")
ref <- t(SC$X)


#annotations
cell_type <- SC$obs$cell_type
cell_type <- factor(cell_type)
CellID <- factor(colnames(ref))
cluster<- data.frame(cellname = CellID, celltype= cell_type)
typeanno <- cluster$celltype
names(typeanno) <- cluster$cellname
typeanno <- as.factor(typeanno)

#spatialdata

st_path <- paste("./simulator_outputs/simulated_gene_expression",i,"_average.h5ad", sep ="")
ST <- read_h5ad(st_path)

spots <-as.sparse(t(ST$X))
class(spots) <- StripAttr(class(spots))
spots <- round(spots)

row_ind <- which(rowSums(spots) > 0)
col_ind <- which(colSums(spots) > 0)
spots <- spots[row_ind,]
spots <- spots[, col_ind]

pos <- ST$obsm$spatial[col_ind,]

# row.names(pos)<-colnames(st_adata)
colnames(pos) <- c('x','y')
coords <- as.data.frame(pos)
row.names(coords)<-colnames(spots)
#get the overlap genes
overlap_gene <- intersect(rownames(spots), rownames(ref))
ref <- ref[overlap_gene,]

spots <- spots[overlap_gene,]

#calculate the nUMI and nUMI_spot
nUMI <- colSums(ref)
names(nUMI) <- colnames(ref)
nUMI_spot <- colSums(spots)
names(nUMI_spot) <- colnames(spots)

#preprocess the input data
class(ref) <- StripAttr(class(ref))
class(spots) <- StripAttr(class(spots))
processed_input<-SONAR.preprocess(sc_count=ref,sc_cell_type=typeanno,sc_nUMI=nUMI,sp_coords=coords,sp_count=spots,sp_nUMI=nUMI_spot,cores=8,type_min_cell = 0 , spot_min_UMI = 0)

#deliver the preprocessed data to SONAR
trans_data<-SONAR.deliver(processed_data=processed_input,path=code_path)

#define the bandwidth, default is 1.2 times minimal distance
temp<-dist(coords)
temp<-Matrix::Matrix(temp)
temp[temp==0] <- NA
mindist <- min(temp,na.rm = T)
h <- 1.2*mindist

#start deconvolution
SONAR.deconvolute(fname = paste0(code_path,"SONAR_main.m"),path=code_path,h,wait = TRUE)
## if return 0, deconvolution has done

SONAR.results <- readMat(paste(code_path,"SONAR_results.mat",sep = ""), drop = FALSE)
SONAR.results <- SONAR.results$JIE
u <- fread(paste0(code_path,"u.txt"))
u[,1] <- NULL
colnames(SONAR.results) <- colnames(u)
spot_name <- read.table(file=paste0(code_path,"coord.txt"),sep=",")
rownames(SONAR.results) <- rownames(spot_name)

setwd("~/Documents/Spatial/Datasets/Spatial_Deconvolution/DLPFC/deconvolution_outputs/SONAR_dlpfc_results")
path <- paste(i,".csv" , sep = "")
final <- SONAR.results[ , sort(colnames(SONAR.results))]
write.csv(final,file=path)
#Complete! SONAR.results is the final results

```

**Comment by lzygenomics** [comment]
Thank you so much for your patient and careful review. I'm sorry for any inconvenience caused. The example we provided has exactly 12 cell types and a limited number of spots. I suspect that the path didn't work when R called Matlab, so the displayed result is still from the example data. I'll double-check the code during work hours. Thank you again!

**Comment by lzygenomics** [comment]
Thank you for your patience! 

【Cause of this issue】 I noticed from the format of the path in your code is based on Linux, but the path setting in my previous SONAR_main.m file was based on the path format for Windows, so it may not have correctly activate the path, leading to R gathers the results from the example data instead, which caused you trouble. 

【Solution】 I have updated the code related to the path in SONAR_main.m, and you can download the updated file directly and replace the original SONAR_main.m file. I apologize for any inconvenience caused and Please don't hesitate to contact us if you have any further questions~

Good Luck!

**Comment by Aditya-Venkat** [comment]
Hello,

We are running into the following line of error having to do with the path, while running the updated code : 
File:
/data/ajita/Spatial/benchmarking_spatial_deconvolution/SONAR/core-code/SONAR_main.m
Line: 77 Column: 1
Invalid use of operator.

However, we were able to extract the correct output file, in terms of the number of spots and cells by saving the output as a csv file and adding an extra throwaway row on top:

`csvwrite("final.csv", JIE,1,0);`

So our issue has been resolved.

We thank you for your immediate response and effort to help us out.

Best of luck!
